src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,312 @@
+            _listener.SetMeasurementEventCallback<decimal>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+        }
+
+
+
+        public AggregationManager Include(string meterName)
+        {
+            Include(i => i.Meter.Name == meterName);
+            return this;
+        }
+
+        public AggregationManager Include(string meterName, string instrumentName)
+        {
+            Include(i => i.Meter.Name == meterName && i.Name == instrumentName);
+            return this;
+        }
+
+        public AggregationManager Include(Predicate<Instrument> instrumentFilter)
+        {
+            lock (this)
+            {
+                _instrumentConfigFuncs.Add(instrumentFilter);
+            }
+            return this;
+        }
+
+        public AggregationManager SetCollectionPeriod(TimeSpan collectionPeriod)
+        {
+            lock (this)
+            {
+                _collectionPeriod = collectionPeriod;
+            }
+            return this;
+        }
+
+        public void Start()
+        {
+            // if already started or already stopped we can't be started again
+            if (_collectThread != null || _cts.IsCancellationRequested)
+            {
+                // correct usage from internal code should never get here
+                throw new InvalidOperationException("Start can only be called once");
+            }
+
+            if (_collectionPeriod.TotalSeconds < MinCollectionTime)
+            {
+                // correct usage from internal code should never get here
+                throw new InvalidOperationException($"CollectionPeriod must be >= {MinCollectionTime} sec");
+            }
+
+            // This explicitly uses a Thread and not a Task so that metrics still work
+            // even when an app is experiencing thread-pool starvation. Although we
+            // can't make in-proc metrics robust to everything, this is a common enough
+            // problem in .NET apps that it feels worthwhile to take the precaution.
+            _collectThread = new Thread(() => CollectWorker(_cts.Token));
+            _collectThread.Start();
+
+            _listener.Start();
+            _initialInstrumentEnumerationComplete();
+        }
+
+        private void CollectWorker(CancellationToken cancelToken)
+        {
+            try
+            {
+                double collectionIntervalSecs = -1;
+                lock (this)
+                {
+                    collectionIntervalSecs = _collectionPeriod.TotalSeconds;
+                }
+                if (collectionIntervalSecs < MinCollectionTime)
+                {
+                    // correct usage from internal code should never get here
+                    throw new InvalidOperationException($"_collectionPeriod must be >= {MinCollectionTime} sec");
+                }
+
+                DateTime startTime = DateTime.UtcNow;
```

Comment:
>startTime [](http://example.com/codeflow?start=25&length=9)

would Stopwatch work better here?

Reply:
At the moment the EventSource reports a DateTime for the start of each metric collection interval. It appears difficult to preserve that behavior with Stopwatch alone because Stopwatch doesn't expose its start time. The best I see there would be to capture both DateTime.Now and start a stopwatch, then assume there neglible time passed in between the two measurements. It doesn't appear simpler but I may also be missing what you had in mind.

Reply:
if you think it is not simplifying anything then ignore this comment. I was just raising the idea.